src/openslide-vendor-dicom.c

```diff
@@ -711,6 +726,76 @@ static void add_properties(openslide_t *osr, struct dicom_level *level0) {
   }
 }
 
+static bool quickhash_level_pixels(struct _openslide_hash *hash,
+                                   struct dicom_level *level) {
+  double d;
+  if (!get_tag_decimal_str(level->file->metadata, NumberOfFrames, 0, &d)) {
+    return false;
+  }
+  uint32_t num_frames = d;
+
+  int64_t total_bytes = 0;
+  for (uint32_t i = 0; i < num_frames; i++) {
+    DcmFrame *frame = dcm_filehandle_read_frame(NULL,
+                                                level->file->filehandle,
+                                                i + 1);
+    if (frame == NULL) {
+      return false;
+    }
+
+    total_bytes += dcm_frame_get_length(frame);
+    if (total_bytes > (5 << 20)) {
+      return false;
+    }
+
+    _openslide_hash_data(hash,
+                         dcm_frame_get_value(frame),
+                         dcm_frame_get_length(frame));
+
+    dcm_frame_destroy(frame);
+  }
+
+  return true;
+}
+
+static void hash_int64(struct _openslide_hash *hash, int64_t i) {
+  // always hash in network byte order
+  int64_t i_be = GINT64_TO_BE(i);
+  _openslide_hash_data(hash, &i_be, sizeof(i_be));
+}
+
+// the required string-valued DICOM tags
+static const char *quickhash_level_metadata_tags[] = {
+  StudyDate,
+  StudyTime,
+  PatientID,
+  StudyID,
+  SeriesNumber,
+  ReferringPhysicianName,
+  PatientName,
+  PatientBirthDate,
+  PatientSex,
+  PatientOrientation,
```

Comment:
Hmm.  Should we include the content/acquisition date/time in addition to (or in lieu of) the study ones?  If we're including the Patient ID, should we include the Issuer of Patient ID?  Some of the patient and physician metadata feels secondary; should we remove it?  Should we include some UIDs?

@hackermd and @dclunie, can you provide any guidance on a suitable set of attributes here?  We're looking for attributes which, taken together, will uniquely identify a DICOM slide, despite any plausible misbehavior of the operator or software that generated it.

For many slide formats, the available metadata fields can't uniquely identify a slide, so OpenSlide also hashes the compressed pixel data from the lowest-resolution level.  Is that necessary for DICOM?  If there's a combination of metadata attributes which guarantee uniqueness &mdash; in particular, if UID collisions don't happen in the wild &mdash; it'd be great to skip the content hash.

Reply:
@bgilbert, yes DICOM UIDs would ideally be the most reliable identifiers of unique entities, such as one scan of a slide (recognizing that each scan may result in multiple DICOM image instances (files) including pyramid layers, overview and label).

First, to state the obvious, if you were going to use the pixel data hash of just one layer of the pyramid as a proxy for the set of files that are one scan of the slide, then the DICOM [SOP Instance UID](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.12.html#sect_C.12.1) of the file (instance) containing that same layer would be an equally suitable proxy, since it should be already globally unique. DICOM implementers and toolkits are generally pretty disciplined about creating well-formed and unique UIDs for this purpose and it is always required to be present.

(F)TL;DNR:

More generally, a discussion of which UID nominally identifies a particular entity and which to select for your use case depends on whether you want to identify a single DICOM "file" (layer of a pyramid, a label, or an overview (macro) file), in which case SOP Instance UID (in that file) should be sufficient, or you want to identify the set of DICOM files that constitute a a single scan (acquisition) of a slide, in which Acquisition UID would work, if present (_vide infra_), assuming that the same acquisition was not stored multiple times (e.g., compressed differently) in DICOM format, or you also wanted to distinguish the same pixel data content in different formats (e.g., want to distinguish SVS from DICOM for same slide).

Traditionally, the general purpose [Series Instance UID](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.3.html#sect_C.7.3.1) was thought to have been sufficient for identifying the set of files (instances), but that has not always been the case in practice. During somewhat contrived testing scenarios like early [WG26 Connectathons](https://doi.org/10.4103/jpi.jpi_1_18) we saw some implementers reusing the same Series Instance UID for multiple successive scans of the same slide, and you may encounter this in old test objects such as at ftp://medical.nema.org/MEDICAL/Dicom/DataSets/WG26/.

So, because it is theoretically possible for a creator to put more than one scan of a slide in the same series, in [CP 2135](https://dicom.nema.org/medical/dicom/Final/cp2135_ft_addAcquisitionEntityAndUID.pdf) we recently (2022/03/31) added the [Acquisition UID](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.10.html#sect_C.7.10.1) and the [Pyramid UID](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.11.html#sect_C.7.11.1) to try to address these sorts of problems going forward. Acquisition UID would seem to fit your use case if you are looking for an identifier for the set of files constituting one scan of a slide such as might otherwise be stored in one SVS or TIFF file with multiple IFDs.

Since Acquisition UID (and Pyramid UID) are recent additions (and optional at the attribute or module level) you may not find many vendors producing them. For testing you should be able to find some images with them in [recent](https://github.com/ImagingDataCommons/idc-wsi-conversion/tree/main#Reconversion) SM objects in the [Imaging Data Commons](https://portal.imaging.datacommons.cancer.gov/explore/filters/?Modality=SM) though, since I included support for generating them in [my converter](https://www.dclunie.com/pixelmed/software/javadoc/com/pixelmed/convert/TIFFToDicom.html).

I might try using a unique key that was a combination of Acquisition UID and Series Instance UID, since that would cover the case when Acquisition UID is missing and the case when a derived (set of) images, such as re-downsampled or image processed or newly lossy compressed set, which might have the same Acquisition UID but ideally would have a different Series Instance UID. I might even be tempted to include several other attributes that might disambiguate original vs. derived (sets of) images if someone was foolish enough to put them in the same series, such as Instance Creation Date and Time (also optional), even Image Type, although if you are going to include a hash of the top layer pixel data or use SOP Instance UID of the top layer instance as a proxy anyway, this might not be necessary.

If you do use SOP Instance UID, you don't need to include the higher level entity UIDs, since the scope of uniqueness of a UID is supposed to be global, not hierarchically dependent on parent entities, and child entities are not supposed to have multiple parents in the study/series/instance hierarchy.

Not sure that a pixel data hash alone is sufficient though, since if someone sends the same scan twice but with different UIDs for some reason, or generates synthetic pixel data for testing, your hash of the pixel data at the top level of the pyramid may identify scans that you think are the same but are not from the same files. One way to disambiguate these is to include the SOP Instance UID of the DICOM instance (file) that the pixel data is extracted from in the hash value.

We sometimes see this sort of thing in radiology PACS where people think a hash of the pixel data is unique, but it is not because the pixel data contains a test pattern (like a SMPTE pattern) that may have been generated by different sources and stored with different identifying metadata. So while the hash uniquely identifies the pixel data, it does not uniquely identify the pixeldata+metadata.

Reply:
Ah, interesting!

I added the tags you suggest, so the slide hash is now formed from a combination of (if present):

```
  AcquisitionUID,
  InstanceCreationDate,
  InstanceCreationTime,
  PatientBirthDate,
  PatientID,
  PatientName,
  PatientOrientation,
  PatientSex,
  PyramidUID,
  ReferringPhysicianName,
  SeriesInstanceUID,
  SeriesNumber,
  SOPInstanceUID,
  StudyDate,
  StudyID,
  StudyTime,
  Columns,
  Rows,
  TotalPixelMatrixColumns,
  TotalPixelMatrixRows,
```

Plus the hash of the compressed pixel data for the smallest pyramid level, provided it's less than 5mb.

Reply:
Thanks for the writeup @dclunie; that was very helpful!

It sounds as though the UIDs are sufficiently reliable that we can avoid hashing any pixel data, and we shouldn't need to hash a long list of attributes in order to get "enough uniqueness".  I'd particularly prefer to avoid hashing pixel data because the lowest-resolution level may not be especially small.

Since the Acquisition UID refers to a physical event rather than a particular set of output, I don't think it's helpful here.  quickhash-1 is not an integrity check, but intentional changes to slide data or metadata should change the quickhash value.  The Series Instance UID seems more useful.  I expect OpenSlide will always treat a series as indivisible; if a series e.g. has both original and derived data, OpenSlide would either ignore the extra instances or refuse to open the slide.  Thus the quickhash shouldn't need to distinguish between instances in a series.

I'm considering two options:

1. Define the quickhash to include solely the Series Instance UID (or the set of all such UIDs, once we support multi-series Z-stacks).  I have some concerns though:

   - In practice, do we need to care about the Series Instance UID collisions you mentioned?  It sounds as though those wouldn't happen under real-world conditions.

   - Is it possible for a well-behaved third-party tool to modify a series without changing the Series Instance UID?

2. Define the quickhash to include all SOP Instance UIDs.  Since the Series Instance UID is now redundant, ignore it.  This definition is more robust to alteration of the series, but possibly _too_ robust; a notional "slide" is still the same slide if someone deletes a file.  And it feels odd to avoid the UID that most directly matches our intentions.

   We'd hash all SOP Instance UIDs, not just the top (or bottom) of the pyramid, because the latter is sensitive to OpenSlide's construction of the pyramid.  We might need to change the construction logic if it turns out that we mishandle complex series.

@dclunie, does one of these options seem preferable?  Any concerns?